### ISSUE TITLE
[HOLD for payment 2024-12-30] [Search v2.5] Update grammar filters

### ISSUE THREAD

**Issue by luacmartins** [Expensify/App#51966]
Coming from [this comment](https://github.com/Expensify/App/pull/51633#discussion_r1827793048), we want to update the filter grammar to use kebab case instead of camel case so it's a bit more friendly to users. Here are the filters that need to be updated:

- `cardID` to `card`
- `taxRate` to `tax-rate`
- `expenseType` to `expense-type`
- `sortBy` to `sort-by`
- `sortOrder` to `sort-order`
- `reportID` to `reportid`
- `policyID` to `workspace`

<details><summary>Issue Owner</summary>Current Issue Owner: @JmillsExpensify</details>

**Comment by 289Adam289** [comment]
Hi! While working on the search parser, I noticed a feature in Peggy that could be beneficial. What's your opinion on making grammar keys case-insensitive, allowing `from` and `From` to function identically? It should be easy to implement by adding letter `i` at the end of a literal in the grammar e.g. `"from"` will become `"from"i`. I think this feature would be useful and make the Search Router easier to use.

**Comment by luacmartins** [comment]
Sounds like a nice improvement, let's do it.!

**Comment by luacmartins** [comment]
Not a priority

**Comment by luacmartins** [comment]
Working on it this week

**Comment by 289Adam289** [comment]
Sure, I'd like to work on that

**Comment by luacmartins** [comment]
Nice! All yours

**Comment by 289Adam289** [comment]
Hi,
I have now been working on this task a bit, and I want to share some thoughts.
There are four sources of queries that need to be parsed in Search:
1. User input
2. Query from URL
3. Query from saved search
4. Query created by the app via advanced filters

The biggest problem with renaming in the UI layer only is that we have one parser to parse all types of queries. It is impossible to allow users only to use new filter keys without doing very big changes to whole Search codebase. We cannot rename filter keys all over the app as there are old URLs and old saved searches that would be incompatible with new names, not to mention changing hashes and payload sent to the backend.
However, I believe I've found a solution that should work:
I changed the parser to allow for both old and new keys, e.g., `taxRate` and `tax-rate`, but we will always show the new, more approachable names to the user. This approach also gives us flexibility and allows for simpler changes in the future. The only downside of this approach is that technically users will still be able to type in old names into input, but these names will be swapped for new ones after the search. However if users won't know old keys, then they should not type them in.

cc @luacmartins for thoughts

**Comment by luacmartins** [comment]
That's exactly what I was thinking. Changing the keys in a deep level would break a lot of functionality, so we need to do it as you mentioned.

**Comment by garrettmknight** [comment]
$250 approved for @hoangzinh

**Comment by JmillsExpensify** [comment]
Payment summary above is correct.

**Comment by JmillsExpensify** [comment]
@luacmartins btw, I assume that we're handling the regression tests centrally? Or should we update/create them with this issue?

**Comment by luacmartins** [comment]
Hmm I think in this case we can do it as part of this issue since the grammar was introduced in v2.2 and we should already have tests for those

**Comment by luacmartins** [comment]
I think we're still waiting on updated test cases https://github.com/Expensify/App/issues/51966#issuecomment-2565676990

**Comment by JmillsExpensify** [comment]
@hoangzinh Mind recommending/pasting the test issues please, per the [BZ checklist](https://github.com/Expensify/App/issues/51966#issuecomment-2560289850)?

**Comment by JmillsExpensify** [comment]
Apologies. No need to double check the current test rail, I am mainly looking for a summary of all the testing steps from the PRs? Then I can work with QA on what exists and doesn't.

**Comment by JmillsExpensify** [comment]
Awesome, much appreciated!

**Comment by JmillsExpensify** [comment]
I'll get the regression test request created, and then we can see what QA thinks about where this best fits. Thanks for the posts and the links. As for payment, summary is below.

 - Contributor: $250 to @hoangzinh via New Expensify.

**Comment by JmillsExpensify** [comment]
I also believe that payment has already been made/approved, right? Basically per [this comment](https://github.com/Expensify/App/issues/51966#issuecomment-2565643181).

**Comment by JmillsExpensify** [comment]
Nice, I'll close then when I create the regression test.

**Comment by luacmartins** [comment]
@JmillsExpensify are we good to close?

**Comment by luacmartins** [comment]
Bump @JmillsExpensify

**Comment by JmillsExpensify** [comment]
Reached out to QA on best next steps for this one.

**Comment by JmillsExpensify** [comment]
Regression test request created, so closing this out.
